book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,437 @@
+
+```rust
+pub struct Chain {
+    ..
+    value_pool: ValueBalance<NonNegative>,
+}
+```
+
+### Update `update_chain_state_with()` for `Chain` struct to calculate Chain.value_pool
+
+- Location: `zebra-state/src/service/non_finalized_state/chain.rs`
+
+```rust
+fn update_chain_state_with(&mut self, prepared: &PreparedBlock) {        
+    ...
+    self.value_pool += prepared.block_value_balance;
+    ...
+}
+```
+
+### Check pool balance consensus rules
+
+#### Check that no pool is negative
+
+- Do the check in `commit_block` and `commit_new_chain`, located in `zebra-state/src/service/non_finalized_state.rs`. Both methods will need new a new argument with the value pool saved in the disk.
+- Pass this value to a new method `check_value_pools()` that will make the consensus rules check.
+
+```rust
+pub fn commit_block(
+    &mut self,
+    prepared: PreparedBlock,
+    finalized_tip_history_tree: &HistoryTree,
+    finalized_value_pool: ValueBalance<NonNegative>,
+) -> Result<(), ValidateContextError> {
+    ..
+    check_value_pools(finalized_value_pools, prepared.value_balance)?;
+    ..
+}
+
+pub fn commit_new_chain(
+    &mut self,
+    prepared: PreparedBlock,
+    finalized_tip_history_tree: HistoryTree,
+    finalized_value_pool: ValueBalance<NonNegative>,
+) -> Result<(), ValidateContextError> {
+    ..
+    check_value_pools(finalized_value_pools, prepared.value_balance)?;
+    ..
+}
+
+fn check_value_pools(
+    finalized_value_pools: ValueBalance<NonNegative>,
+    block_value_balance: ValueBalance<NegativeAllowed>
+) -> Result<()> {
+    if finalized_value_pools + block_value_balance < 0 {
+        Err("Value pool can't be negative");
+    }
+    Ok(())
+}
+```
+
+##### Consensus rule
+
+The chain value pool balance rules apply to Block transactions, but they are optional for Mempool transactions:
+> Nodes MAY relay transactions even if one or more of them cannot be mined due to the aforementioned restriction.
+
+https://zips.z.cash/zip-0209#specification
+
+Since Zebra does chain value pool balance validation in the state, we want to skip verifying the speculative chain balance of Mempool transactions.
+
+### Changes to finalized state
+
+The state service is what will call `commit_block()` and `commit_new_chain()`. We need to pass the value pool from the disk into this functions.
+
+```
+self.mem
+    .commit_block(prepared, self.disk.history_tree(), self.disk.get_pool())?;
+..
+self.mem
+    .commit_new_chain(prepared, self.disk.history_tree(), self.disk.get_pool())?;
+```
+
+We now detail what is needed in order to have the `get_pool()` method available.
+
+#### Serialization of `ValueBalance<C>`
+
+In order to save `ValueBalance` into the disk database we must implement `IntoDisk` and `FromDisk` for `ValueBalance` and for `Amount`:
+
+```rust
+
+impl IntoDisk for ValueBalance<C> {
+    ..
+}
+
+impl FromDisk for ValueBalance<C> {
+    ..
+}
```

Comment:
@oxarbitrage how do you want to serialize the 4 value pool amounts into bytes?

Can you fill in the details here?

Reply:
I was thinking on something as https://github.com/ZcashFoundation/zebra/pull/2430/commits/934847272c6d27b4491a4cae93fa5dcd10ca3055 however not sure if it will be the best.